File: backend/app/api/services/financas/encargos.py
"""Encargos por atraso de boleto: multa (uma vez) + juros de mora (por dia).

Padrão brasileiro: ao vencer, incide uma **multa** percentual única sobre o
valor e **juros de mora** mensais cobrados pró-rata por dia de atraso (juros
simples). Ex.: "multa de 2% + juros de 1% ao mês" → no 10º dia de atraso de um
boleto de R$100: multa 2,00 + juros 100·1%·(10/30) = 0,33 → encargos 2,33.
"""
from __future__ import annotations

from datetime import date
from decimal import Decimal, ROUND_HALF_UP
from typing import Optional

CENTAVO = Decimal("0.01")


def calcular_encargos(
    valor: Decimal,
    vencimento: Optional[date],
    multa_percentual: Optional[Decimal],
    juros_mensal_percentual: Optional[Decimal],
    referencia: date,
) -> Decimal:
    """Multa + juros projetados sobre ``valor`` até ``referencia``.

    Retorna 0 se não há vencimento, não está atrasado, ou não há encargos.
    Juros mensais são convertidos linearmente pra diário (÷30)."""
    if vencimento is None or referencia <= vencimento:
        return Decimal("0")
    dias = (referencia - vencimento).days
    valor = Decimal(valor)
    total = Decimal("0")
    if multa_percentual:
        total += valor * Decimal(multa_percentual) / Decimal("100")
    if juros_mensal_percentual:
        total += (
            valor
            * Decimal(juros_mensal_percentual) / Decimal("100")
            * Decimal(dias) / Decimal("30")
        )
    return total.quantize(CENTAVO, rounding=ROUND_HALF_UP)

File: frontend/lib/encargos.ts
/**
 * Encargos por atraso de boleto: multa (uma vez) + juros de mora (por dia).
 * Espelha `app/api/services/financas/encargos.py` no backend — a conta na hora
 * de pagar usa a mesma fórmula e a mesma data, então o valor mostrado bate com
 * o que é efetivado.
 */

function paraData(iso: string): Date {
  return new Date(iso.slice(0, 10) + 'T00:00:00');
}

function num(v: string | number | null | undefined): number {
  const n = typeof v === 'string' ? Number(v) : v;
  return Number.isFinite(n as number) ? (n as number) : 0;
}

export interface Encargos {
  dias: number;
  multa: number;
  juros: number;
  total: number;
}

/**
 * Multa + juros sobre `valor` até `referencia` (ISO YYYY-MM-DD). Tudo zero se
 * não há vencimento, está no prazo, ou não há encargos. Juros mensais viram
 * diários linearmente (÷30).
 */
export function calcularEncargos(
  valor: string | number,
  vencimento: string | null | undefined,
  multaPercentual: string | number | null | undefined,
  jurosMensalPercentual: string | number | null | undefined,
  referencia: string,
): Encargos {
  const zero: Encargos = { dias: 0, multa: 0, juros: 0, total: 0 };
  if (!vencimento) return zero;
  const venc = paraData(vencimento);
  const ref = paraData(referencia);
  const dias = Math.round((ref.getTime() - venc.getTime()) / 86_400_000);
  if (dias <= 0) return zero;

  const base = num(valor);
  const multaPct = num(multaPercentual);
  const jurosPct = num(jurosMensalPercentual);
  const multa = (base * multaPct) / 100;
  const juros = ((base * jurosPct) / 100) * (dias / 30);
  const round = (n: number) => Math.round(n * 100) / 100;
  const m = round(multa);
  const j = round(juros);
  return { dias, multa: m, juros: j, total: round(m + j) };
}
